models/sage_intacct__balance_sheet.sql

```diff
@@ -1,9 +1,56 @@
 with general_ledger_by_period as (
-    select *
+    select 
+        period_first_day,
+        account_no,
+        account_title,
+        account_type,
+        book_id,
+        category,
+        classification,
+        currency,
+        entry_state,
+        period_ending_amount
     from {{ ref('sage_intacct__general_ledger_by_period') }}
     where account_type = 'balancesheet'
 ), 
 
+general_ledger_by_period_retained_earnings_tmp as (
+    select
+        period_first_day,
+        '999999' as account_no,
+        'Retained Earnings' as account_title,
+        'balancesheet' as account_type,
+        'ACCRUAL' as book_id,
+        'Retained Earnings' as category,
+        'Equity' as classification,
+        'currency' as currency,
+        'P' as entry_state,
+        sum(period_net_amount) as period_net_amount
+    from {{ ref('sage_intacct__general_ledger_by_period') }}
+    where account_type = 'incomestatement'
+    group by 1
+),
+
+general_ledger_by_period_retained_earnings as (
+    select
+        period_first_day,
+        '999999' as account_no,
+        'Retained Earnings' as account_title,
+        'balancesheet' as account_type,
+        'ACCRUAL' as book_id,
+        'Retained Earnings' as category,
+        'Equity' as classification,
+        'currency' as currency,
+        'P' as entry_state,
+        sum(period_net_amount) over (
+            order by period_first_day
+            rows between unbounded preceding and current row
+        ) as amount
+    from general_ledger_by_period_retained_earnings_tmp
+    order by period_first_day
```

Comment:
Finally, why do we have "P" as the set value for entry_state?

Reply:
One last question is around the tmp cte? What are we attempting to do with these two ctes?

Reply:
Thanks for the comments!
1. You make a good point of creating the retained earnings bucket in the sage_intacct__general_ledger_by_period instead, and this will allow the final balance sheet to remain straightforward. I think it makes sense to move it.
2. Currency and entry_state: wasn't sure what to have as the default there. I didn't want to hard code, for example, USD as currency in case a customer isn't using USD. Lmk if you have an opinion on what to use as defaults
3. The 2 new ctes are to create a retained earnings category which takes the cumulative period_amount. When it rolls up to Equity, the balance sheet then rounds out.

Reply:
> You make a good point of creating the retained earnings bucket in the sage_intacct__general_ledger_by_period instead, and this will allow the final balance sheet to remain straightforward. I think it makes sense to move it.

- Yeah let's move it

> Currency and entry_state: wasn't sure what to have as the default there. I didn't want to hard code, for example, USD as currency in case a customer isn't using USD. Lmk if you have an opinion on what to use as defaults

- Let's confirm what the appropriate entry_state should be here, if the customer chose `P` then we can stick with that. However, I think it would be better to leave currency as null. I agree that hard coding `USD` would not be appropriate, but maybe a better alternative would to just leave null and let the customer backfill the currency if they would like.

> The 2 new ctes are to create a retained earnings category which takes the cumulative period_amount. When it rolls up to Equity, the balance sheet then rounds out.

- Makes sense, but should we maybe have this exist as it's own intermediate model instead of being nested in another model? This way it will be better broken out?